include/dali/core/tensor_layout.h

```diff
@@ -0,0 +1,475 @@
+    assert(start_dim >= 0 && start_dim <= ndim());
+    if (n < 0)
+      n = ndim() - start_dim;
+    assert(start_dim + n <= ndim());
+    return TensorLayout(begin() + start_dim, n);
+  }
+
+  DALI_HOST_DEV
+  constexpr TensorLayout first(int n) const noexcept {
+    assert(n <= ndim());
+    return TensorLayout(begin(), n);
+  }
+
+  DALI_HOST_DEV
+  constexpr TensorLayout last(int n) const noexcept {
+    assert(n <= ndim());
+    return TensorLayout(end() - n, n);
+  }
+
+  static constexpr int max_ndim = 15;
+
+ private:
+  /** @brief Stores the dimension descriptions as a null-terminated string
+   *
+   * Last character in the string is aliased with max_ndim - length.
+   * If length == max_ndim, then 0 is stored and aliased as '\0' terminator,
+   * allowing 1 extra character to be stored.
+   */
+  char data_[max_ndim + 1] = { 0 };
+  DALI_HOST_DEV
+  void set_size(int n) {
+    assert(n >= 0 && n <= max_ndim);
+    data_[max_ndim] = max_ndim - n;
+  }
+  DALI_HOST_DEV
+  friend constexpr TensorLayout operator+(const TensorLayout &a, const TensorLayout &b);
+};
+
+static_assert(sizeof(TensorLayout) == 16, "Tensor layout size should be exactly 16B");
+
+DALI_HOST_DEV
+constexpr TensorLayout operator+(const TensorLayout &a, const TensorLayout &b) {
+  assert(a.size() + b.size() < TensorLayout::max_ndim);
+  TensorLayout result = a;
+  int i = 0, j = 0;
+  for (i = result.ndim(), j = 0; i < TensorLayout::max_ndim && j < b.ndim(); i++, j++)
+    result[i] = b[j];
+  result[i] = 0;
+
+  /* Cannot use
+   *   result.set_size(i);
+   * because it's not constexpr
+   */
+  result.data_[TensorLayout::max_ndim] = TensorLayout::max_ndim - i;
+  return result;
+}
+
+
+DALI_HOST_DEV
+constexpr TensorLayout TensorLayout::skip(char dim_name) const noexcept {
+  int i = find(dim_name);
+  if (i < 0)
+    return *this;
+  return first(i) + sub(i+1);
+}
+
+#define DEFINE_TENSOR_LAYOUT_COMPARISON(op)                             \
+inline bool operator op(const TensorLayout &tl, const std::string &s) { \
+  return std::strcmp(tl.c_str(), s.c_str()) op 0;                       \
+}                                                                       \
+DALI_HOST_DEV                                                           \
+constexpr bool operator op(const TensorLayout &tl, const char *s) {     \
+  return tl.compare(s) op 0;                                            \
+}                                                                       \
+inline bool operator op(const std::string &s, const TensorLayout &tl) { \
+  return std::strcmp(s.c_str(), tl.c_str()) op 0;                       \
+}                                                                       \
+DALI_HOST_DEV                                                           \
+constexpr bool operator op(const char *s, const TensorLayout &tl) {     \
+  return TensorLayout(s).compare(tl) op 0;                              \
+}
+
+DEFINE_TENSOR_LAYOUT_COMPARISON(<)  // NOLINT
```

Comment:
I'm not sure what is the meaning of one layout being `<` then the other in a sense that it is implemented here? Is there a use case for it? If no, I would vote for `==` and `!=` only.

Reply:
Operator `<` is needed for set/map/... And certainly I don't like to have just one of them defined.
Apart form that, I want this object to be as similar to a string as possible.